### PR TITLE
Switch from `winapi` to `windows-sys`

### PR DESCRIPTION
**Pull request by CraftSpider**
This is a rebase of #494, as it was mentioned in #537 that there was potential interest in such a thing.

### CODE REVIEW

**Comment by CraftSpider** [comment]
Currently, this avoid actually depending on `windows-bindgen` at build-time by simply using a file generated with the arguments in `bindings.txt`. This is similar to how `std` does it, and avoids issues with `as-if-std`. While similar to the previous method, this way the windows system APIs are generated entirely by the same system as `windows-sys`, and are based on the official API metadata.

**Comment by ChrisDenton** [comment]
If we're using generated bindings then I don't think we need `verify-windows-sys`. There's no need to hand-write the bindings when we can use the generated bindings directly.

File: src/backtrace/dbghelp32.rs [hunk 1]
```diff
@@ -176,7 +176,7 @@ pub unsafe fn trace(cb: &mut dyn FnMut(&super::Frame) -> bool) {
             };
 
             while dbghelp.StackWalk64()(
-                image as DWORD,
+                image as _,
```
Comment (ChrisDenton) [inline src/backtrace/dbghelp32.rs:179]: Please use an explicit type here.

**Review by ChrisDenton — COMMENTED** [review]
Looks good to me. Just one nit.

File: Cargo.toml [hunk 2]
```diff
@@ -20,10 +20,10 @@ rust-version = "1.65.0"
 [workspace]
 members = ['crates/cpp_smoke_test', 'crates/as-if-std']
 exclude = [
-  'crates/without_debuginfo',
-  'crates/macos_frames_test',
-  'crates/line-tables-only',
-  'crates/debuglink',
+    'crates/without_debuginfo',
```
Comment (CraftSpider) [inline Cargo.toml:23]: Reminder to self - revert this change, fix my config to not do this

File: src/dbghelp.rs [hunk 3]
```diff
@@ -12,7 +12,7 @@
 //! actually use the raw definitions in `winapi`, but rather we need to define
 //! the function pointer types ourselves and use that. We don't really want to
 //! be in the business of duplicating winapi, so we have a Cargo feature
-//! `verify-winapi` which asserts that all bindings match those in winapi and
+//! `verify-windows-sys` which asserts that all bindings match those in winapi and
```
Comment (CraftSpider) [inline src/dbghelp.rs:15]: Reminder to self - Remove reference to winapi here

**Comment by CraftSpider** [comment]
> ~If we're using generated bindings then I don't think we need `verify-windows-sys`. There's no need to hand-write the bindings when we can use the generated bindings directly.~
> 
> EDIT: never mind, that won't work because `windows-bindgen` doesn't have a way to generate delay-loaded imports.

It may be a good idea to remove the feature and make the assertion unconditional - it's purely compile-time, and since we are comparing to the auto-generated bindings, we always have access to the original definitions. Either a feature in windows-bindgen to generate `type` for each function, or access to `typeof` in Rust, would let us just use the binding types directly, but alas.

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -144,7 +144,7 @@ jobs:
       shell: bash
     - run: rustup target add aarch64-pc-windows-msvc
     - run: cargo test --no-run --target aarch64-pc-windows-msvc
-    - run: cargo test --no-run --target aarch64-pc-windows-msvc --features verify-winapi
+    - run: cargo test --no-run --target aarch64-pc-windows-msvc --features verify-windows-sys
```
Comment (workingjubilee) [inline .github/workflows/main.yml:147]: ...I have left it as-is until recently, but why do we not simply run this all the time when we know we're on Windows?

File: src/dbghelp.rs [hunk 5]
```diff
@@ -357,7 +303,7 @@ pub fn init() -> Result<Init, ()> {
         }
         debug_assert!(lock != 0);
         let lock = lock as HANDLE;
-        let r = WaitForSingleObjectEx(lock, INFINITE, FALSE);
+        let r = WaitForSingleObjectEx(lock, INFINITE, 0);
```
Comment (workingjubilee) [inline src/dbghelp.rs:306]: I do still prefer being able to see when something is actually only supposed to have two values.
```suggestion
        let r = WaitForSingleObjectEx(lock, INFINITE, false as BOOL);
```

File: src/dbghelp.rs [hunk 6]
```diff
@@ -345,7 +291,7 @@ pub fn init() -> Result<Init, ()> {
         let mut lock = LOCK.load(SeqCst);
         if lock == 0 {
             let name = mutex_name();
-            lock = CreateMutexA(ptr::null_mut(), 0, name.as_ptr().cast::<i8>()) as usize;
+            lock = CreateMutexA(ptr::null_mut(), 0, name.as_ptr()) as usize;
```
Comment (workingjubilee) [inline src/dbghelp.rs:294]:
```suggestion
            lock = CreateMutexA(ptr::null_mut(), false as BOOL, name.as_ptr()) as usize;
```

**Review by workingjubilee — COMMENTED** [review]
Only a pair of nits.

File: src/dbghelp.rs (hunk 5 above)
Comment (CraftSpider) [reply src/dbghelp.rs:306]: Whoops! I tried to change `TRUE` -> `1` back, but I apparently missed the `FALSE` -> `0` changes. I'll fix this (though probably using the windows_sys `FALSE` constant, to be consistent with using `TRUE`)

**Comment by workingjubilee** [comment]
Thanks for rebasing! Finally figured out [what busted CI](https://github.com/rust-lang/backtrace-rs/pull/641) so I'm gonna merge this through finally.

**Review by workingjubilee — APPROVED** [review]

**Comment by kleisauke** [comment]
FWIW, this breaks the tier-3 `thumbv7a-pc-windows-msvc` and `thumbv7a-uwp-windows-msvc` targets. For example, `IMAGE_FILE_MACHINE_ARMNT` is gone now.
https://github.com/rust-lang/backtrace-rs/blob/38d49aa0395c50047ab26e418a0cb30d9955fb64/src/backtrace/dbghelp32.rs#L222

PR https://github.com/rust-lang/rust/pull/112527 could be relevant here.

**Comment by workingjubilee** [comment]
Tier 3 targets are allowed to be broken. If they weren't, I wouldn't be able to even land this PR. If someone wants them to be supported, they need to start maintaining them.  https://doc.rust-lang.org/rustc/target-tier-policy.html

**Comment by workingjubilee** [comment]
And the appropriate venue for this is **not** in the tail end of a PR, but a new issue.

**Comment by kleisauke** [comment]
Sorry, the last time I dealt with a similar issue (see #572) the fix was straightforward, but this is somewhat non-trivial as `windows-sys` lacks platform-specific type definitions for ARM32.

That said, this is a good opportunity to remove this unsupported Microsoft target from my toolchain. If anyone fancies it, they can fix this, but it won't be me this time.
